proposals/accepted/000-ast-parser-libs.rst

```diff
@@ -0,0 +1,344 @@
+
+In `this comment <https://gitlab.haskell.org/ghc/ghc/-/issues/14409#note_506489>`_, it is suggested that factoring out the AST and parser can be a good first step making a more modular in GHC in general.
+This proposal wish to *stay neutral* on the merits of such a future direction, but it would be remiss not to at least highlight it as one possible outcome.
+
+Roadmap
+=======
+
+The project is split into two separate steps: separating the AST, and separating the parser.
+Each step has a method, time estimate, and (most importantly) clear success criteria, including use by downstream projects to ensure value is delivered.
+The intent is thus that they are self-contained, and can be individually funded.
+
+Separate the AST
+----------------
+
+Split library
+~~~~~~~~~~~~~
+
+**Executor**: Haskell Foundation
+
+**Time Estimate:** 1 – 2 Weeks
+
+The first step is just separating data definitions.
+We don't need to worry about code entangling, just data entangling.
+We have already separated those data definitions into modules in the ``Language.Haskell.Syntax.*`` namespace.
+
+Concretely, the work in this step is to:
+
+#. Modify those modules to not import any other modules in ``ghc`` (``GHC.*`` modules).
+
+#. Move those modules to a new separate AST library in the GHC repo.
+
+#. Adjust ``build-depends`` across the repo so ``ghc`` and any other Haskell Package gets those modules from the new library instead, and CI passes.
+
+The timeline for this is pretty short because there exists an easy last-resort way to decouple anything:
+just add another TTG type family.
+This came up with some acrimony in `GHC Issue #21628 <https://gitlab.haskell.org/ghc/ghc/-/issues/21628>`_, discussing whether it was better to try to change GHC's ``FastString`` or abstract over it.
+The purpose of this proposal isn't to relitigate that issue, but because this proposal *is* about resource allocation, something does need to be said on the broader tradeoffs at play
+
+There is no disagreement that as-is, that data type is not suitable for a nice self-contained library. [#faststring-unsuitable]_
+The disagreement is whether TTG should be blocked on reworking ``FastString`` somehow to be better for GHC and non-GHC alike, or whether we should just side-step the issue entirely.
+
+We make no claims about what is better in the long term for GHC, but when reworking ``FastString`` and benchmarking the new algorthms might take **Days to Weeks**, we can side-step the issue with a new ``StringP`` type family "extension point" like the existing ``IdP`` one in **minutes**. [#extension-point]_
+
+Out of a basic fiduciary duty, we thus declare that unless "Plan A" works out almost as quickly, "Plan B" of just introducing another extension point should be used.
```

Comment:
I don't think this is a fiduciary duty - those are duties incurred by virtue of being someone else's financial representative, where you must represent their interests. I'd say "To keep costs under control" or similar

Reply:
Why does this need to be an extension point with a type family, rather than parameterising the AST over some arbitrary type of strings?

Reply:
I think part of the point of TTG is that you just have _one_ parameter to the AST type, and all the other bits hang off that with type families. Otherwise you would have a truly ungodly number of type parameters.

Reply:
Sounds good to me